### PR TITLE
DDF-4658 Adding toggle for spellcheck feature

### PR DESCRIPTION
**Pull request by cantstoptheunk**
#### What does this PR do?
Adds the ability to turn on/off the spellcheck feature in search settings. 
Also adds the ability to turn on/off the spellcheck in the admin settings (turning it off in the admin under Catalog UI Search means that the spellcheck toggle will not even show up inside intrigue)

#### Who is reviewing it? 
@bellcc 
@derekwilhelm 
@andrewkfiedler 
@tbatie 

#### Ask 2 committers to review/merge the PR and tag them here.
@andrewkfiedler
@tbatie

#### How should this be tested?
1. Upload/ingest data
2. Check that you are able to toggle hit on/off for spellcheck in Default Search Settings and in Query Search Settings without errors.
3. Go to admin config under Catalog UI Search and turn off Spellcheck.
4. Go back to intrigue and ensure that the spellcheck option no longer appears in Default Search Settings nor in Query Search Settings

#### What are the relevant tickets?
For GH Issues:
Fixes: #4658 

#### Checklist:
- [ ] Documentation Updated
- [ ] Update / Add Threat Dragon models
- [ ] Update / Add Unit Tests
- [ ] Update / Add Integration Tests

#### Notes on Review Process
Please see [Notes on Review Process](https://codice.atlassian.net/wiki/spaces/DDF/pages/71946981/Pull+Request+Guidelines) for further guidance on requirements for merging and abbreviated reviews. 

#### Review Comment Legend:
- ✏️ (Pencil) This comment is a nitpick or style suggestion, no action required for approval. This comment should provide a suggestion either as an in line code snippet or a gist. 
- ❓ (Question Mark) This comment is to gain a clearer understanding of design or code choices, clarification is required but action may not be necessary for approval.
- ❗ (Exclamation Mark) This comment is critical and requires clarification or action before approval.

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -21,11 +21,15 @@ atlassian-ide-plugin.xml
 platform/solr/**/overlays
 distribution/docs/src/main/resources/content/**/.asciidoctor
 distribution/docs/src/main/resources/content/**/*.png
+*.class
```
Comment (ethantmanns) [inline .gitignore]: Compiled classes should be in the `target/` dir that's excluded below. I don't think we should need to add this 🤔

File: ui/packages/catalog-ui-search/src/main/webapp/component/search-interactions/search-interactions.view.js [hunk 2]
```diff
@@ -50,7 +50,6 @@ module.exports = Marionette.LayoutView.extend({
       this.triggerCloseDropdown
     )
     this.generateSearchFormSelector()
-    this.generateSearchSettings()
```
Comment (samuelechu) [inline ui/packages/catalog-ui-search/src/main/webapp/component/search-interactions/search-interactions.view.js]: why was this deleted?

File: ui/packages/catalog-ui-search/src/main/webapp/component/query-settings/query-settings.view.js [hunk 3]
```diff
@@ -158,6 +162,25 @@ module.exports = plugin(
...
+      const spellcheckView = Marionette.ItemView.extend({
+        template: () => <RadioComponent value={this.model.get('spellcheck')} label="Spellcheck" options={[{
+          label: 'Yes',
+          value: true
+        }, {
+          label: 'No',
+          value: false
+        }]}
+          onChange={(value) => {
+            this.model.set('spellcheck', value)
```
Comment (samuelechu) [inline ui/packages/catalog-ui-search/src/main/webapp/component/query-settings/query-settings.view.js]: Should `change:spellcheck` be added here? https://github.com/codice/ddf/pull/4698/files#diff-1d3ab149b2771a1eaea93cb9f46d7c0bR63

File: distribution/docs/src/main/resources/content/_reference/_tables/catalog.ui.adoc [hunk 4]
```diff
@@ -246,6 +246,13 @@ Example:
 |true
 |true
 
+|Enable Spellcheck
+|spellcheckEnabled
+|Boolean
+|Enable spellcheck for seraches.
```
Comment (ricklarsen) [inline distribution/docs/src/main/resources/content/_reference/_tables/catalog.ui.adoc]: searches?

File: ui/packages/catalog-ui-search/src/main/webapp/component/search-interactions/search-interactions.view.js (hunk 2 above)
Comment (cantstoptheunk) [reply ui/packages/catalog-ui-search/src/main/webapp/component/search-interactions/search-interactions.view.js]: This is to remove the confusing settings option in the searches. It used to exist under the `User another search form` option
Screen Shot 2019-04-16 at 2 09 42 PM

File: ui/packages/catalog-ui-search/src/main/webapp/component/search-interactions/search-interactions.hbs [hunk 5]
```diff
@@ -37,5 +37,4 @@
     <div class="interaction-text">
         Reset
     </div>
-</div>
-<div class="interaction interaction-settings" title="Contains search settings and defaults." data-help="Contains search settings and defaults."></div>
```
[the remark this answers is not in the document]
Comment (cantstoptheunk) [reply ui/packages/catalog-ui-search/src/main/webapp/component/search-interactions/search-interactions.hbs]: See comment below

File: ui/packages/catalog-ui-search/src/main/webapp/component/query-settings/query-settings.view.js [hunk 6]
```diff
@@ -29,8 +29,10 @@ const Common = require('../../js/Common.js')
 const properties = require('../../js/properties.js')
 const plugin = require('plugins/query-settings')
 const announcement = require('../announcement/index.jsx')
-import { InvalidSearchFormMessage } from 'component/announcement/CommonMessages'
 const ResultForm = require('../result-form/result-form.js')
+import { InvalidSearchFormMessage } from 'component/announcement/CommonMessages'
+import * as React from 'react'
+import RadioComponent from '../../react-component/container/input-wrappers/radio'
```
[the remark this answers is not in the document]
Comment (cantstoptheunk) [reply ui/packages/catalog-ui-search/src/main/webapp/component/query-settings/query-settings.view.js]: I believe this is not the case, it should refer to the index.tsx file in the same directory.

File: ui/packages/catalog-ui-search/src/main/webapp/component/input/boolean/input-boolean.view.js [hunk 7]
```diff
@@ -13,12 +13,19 @@
  *
  **/
 
+import InputBoolean from './input-boolean.presentation'
```
[the remark this answers is not in the document]
Comment (cantstoptheunk) [reply ui/packages/catalog-ui-search/src/main/webapp/component/input/boolean/input-boolean.view.js]: I am happy to tag up with you since I am not seeing this problem on my devserver

File: ui/packages/catalog-ui-search/src/main/webapp/component/input/boolean/input-boolean.view.js [hunk 8]
```diff
@@ -13,12 +13,19 @@
...
+import * as React from 'react'
+
 var Marionette = require('marionette')
 var template = require('./input-boolean.hbs')
 var InputView = require('../input.view')
 
 module.exports = InputView.extend({
-  template: template,
+  template(props) {
```
[the remark this answers is not in the document]
Comment (cantstoptheunk) [reply ui/packages/catalog-ui-search/src/main/webapp/component/input/boolean/input-boolean.view.js]: No that's not the case.

File: ui/packages/catalog-ui-search/src/main/webapp/react-component/container/input-wrappers/base/base.tsx [hunk 9]
```diff
@@ -62,6 +62,7 @@ type MyProps = {
   value?: any[]
   values?: object
   enumeration?: any[]
+  radio?: any[]
```
[the remark this answers is not in the document]
Comment (cantstoptheunk) [reply ui/packages/catalog-ui-search/src/main/webapp/react-component/container/input-wrappers/base/base.tsx]: Radio is an optional prop, and the `any[]` means that you can assign any object to it.

File: ui/packages/catalog-ui-search/src/main/webapp/component/input/boolean/input-boolean.view.js (hunk 7 above)
[the remark this answers is not in the document]
Comment (cantstoptheunk) [reply ui/packages/catalog-ui-search/src/main/webapp/component/input/boolean/input-boolean.view.js]: I had to commit the untracked files. The file should now be pushed up.

File: ui/packages/catalog-ui-search/src/main/webapp/react-component/container/input-wrappers/radio/radio.tsx [hunk 10]
```diff
@@ -0,0 +1,36 @@
+/**
```
Comment (willwill96) [inline ui/packages/catalog-ui-search/src/main/webapp/react-component/container/input-wrappers/radio/radio.tsx]: ✏️ Consider using the already existing radio button component: https://github.com/codice/ddf/blob/0adc9b6cc16e30f1c82852c0f0bc87207d02ec20/ui/packages/catalog-ui-search/src/main/webapp/component/input/radio/input-radio.view.js

File: ui/packages/catalog-ui-search/src/main/webapp/component/input/boolean/input-boolean.view.js [hunk 11]
```diff
@@ -13,12 +13,21 @@
  *
  **/
 
+import InputBoolean from './input-boolean.presentation'
+import * as React from 'react'
+
 var Marionette = require('marionette')
 var template = require('./input-boolean.hbs')
```
Comment (willwill96) [inline ui/packages/catalog-ui-search/src/main/webapp/component/input/boolean/input-boolean.view.js]: ✏️ Can remove this line and the template itself

File: ui/packages/catalog-ui-search/src/main/webapp/component/input/boolean/input-boolean.presentation.tsx [hunk 12]
```diff
@@ -0,0 +1,44 @@
+/**
+ * Copyright (c) Codice Foundation
+ *
+ * This is free software: you can redistribute it and/or modify it under the terms of the GNU Lesser
```
Comment (willwill96) [inline ui/packages/catalog-ui-search/src/main/webapp/component/input/boolean/input-boolean.presentation.tsx]: ✏️ This file belongs in the react-component directory

File: ui/packages/catalog-ui-search/src/main/webapp/component/announcement/announcementsError.tsx [hunk 13]
```diff
@@ -0,0 +1,5 @@
+class AnnouncementsError extends React.Component {
```
Comment (willwill96) [inline ui/packages/catalog-ui-search/src/main/webapp/component/announcement/announcementsError.tsx]: ❓ What is the purpose of this component? It doesn't look like you use it.

✏️ It also isn't being exported

File: ui/packages/catalog-ui-search/src/main/webapp/component/input/boolean/input-boolean.view.js [hunk 14]
```diff
@@ -13,12 +13,21 @@
...
 var InputView = require('../input.view')
 
 module.exports = InputView.extend({
-  template: template,
+  template(props) {
```
Comment (willwill96) [inline ui/packages/catalog-ui-search/src/main/webapp/component/input/boolean/input-boolean.view.js]: ❗️ This is going to affect the way that filters looks when selecting a boolean attribute

File: ui/packages/catalog-ui-search/src/main/webapp/component/search-interactions/search-interactions.view.js (hunk 2 above)
[the remark this answers is not in the document]
Comment (cantstoptheunk) [reply ui/packages/catalog-ui-search/src/main/webapp/component/search-interactions/search-interactions.view.js]: Done

File: ui/packages/catalog-ui-search/src/main/webapp/component/announcement/announcementsError.tsx (hunk 13 above)
Comment (cantstoptheunk) [reply ui/packages/catalog-ui-search/src/main/webapp/component/announcement/announcementsError.tsx]: Got rid of it, thank you ser

File: ui/packages/catalog-ui-search/src/main/webapp/component/input/boolean/input-boolean.view.js (hunk 14 above)
Comment (cantstoptheunk) [reply ui/packages/catalog-ui-search/src/main/webapp/component/input/boolean/input-boolean.view.js]: Reverted the change, should be fixed now

File: ui/packages/catalog-ui-search/src/main/webapp/component/input/boolean/input-boolean.view.js (hunk 11 above)
Comment (cantstoptheunk) [reply ui/packages/catalog-ui-search/src/main/webapp/component/input/boolean/input-boolean.view.js]: Reverted the change so it will now be required.

File: ui/packages/catalog-ui-search/src/main/webapp/component/input/boolean/input-boolean.presentation.tsx (hunk 12 above)
Comment (cantstoptheunk) [reply ui/packages/catalog-ui-search/src/main/webapp/component/input/boolean/input-boolean.presentation.tsx]: Done

**Comment by cantstoptheunk** [comment]
build now

File: ui/packages/catalog-ui-search/src/main/webapp/component/query-settings/query-settings.view.js [hunk 15]
```diff
@@ -182,6 +213,7 @@ module.exports = plugin(
     },
     toJSON: function() {
       var federation = this._srcDropdownModel.get('federation')
+      var spellcheck = this.model.get('spellcheck')
```
Comment (willwill96) [inline ui/packages/catalog-ui-search/src/main/webapp/component/query-settings/query-settings.view.js]: ✏️ const

File: ui/packages/catalog-ui-search/src/main/webapp/component/query-settings/query-settings.view.js [hunk 16]
```diff
@@ -201,6 +233,7 @@ module.exports = plugin(
         federation: federation,
         sorts: sorts,
         'detail-level': detailLevel,
+        spellcheck: spellcheck,
```
Comment (willwill96) [inline ui/packages/catalog-ui-search/src/main/webapp/component/query-settings/query-settings.view.js]: ✏️ Could use shorthand. Ex:
```suggestion
        spellcheck,
```

**Comment by cantstoptheunk** [comment]
build now

**Comment by rececoffin** [comment]
Hero in progress

**Comment by cantstoptheunk** [comment]
build now

**Comment by cantstoptheunk** [comment]
build now

**Comment by rececoffin** [comment]
Hero successful: 

- Ingested data, ran query and spellcheck did not run
- Enabled spellcheck in admin console. Was able to toggle in UI
- Disabled spellcheck in admin and toggle was gone. Spellcheck does not work when disabled in admin

File: catalog/core/catalog-core-solr/src/main/java/ddf/catalog/source/solr/SolrMetacardClientImpl.java [hunk 17]
```diff
@@ -271,9 +269,24 @@ public SourceResponse query(QueryRequest request) throws UnsupportedQueryExcepti
...
+    Boolean userSpellcheckChoice = false;
+    if (request.getProperties().get("spellcheck") != null) {
+      userSpellcheckChoice = (Boolean) request.getProperties().get("spellcheck");
+    }
+    return userSpellcheckChoice;
+  }
+
+  private boolean solrSpellcheckHasResults(QueryResponse solrResponse) {
+    return solrResponse.getSpellCheckResponse() != null
+        && solrResponse.getResults().size() == 0
```
Comment (bdeining) [inline catalog/core/catalog-core-solr/src/main/java/ddf/catalog/source/solr/SolrMetacardClientImpl.java]: can getResults be null?

Comment (cantstoptheunk) [reply catalog/core/catalog-core-solr/src/main/java/ddf/catalog/source/solr/SolrMetacardClientImpl.java]: No because if there are no results returned from the query when `clientquery(query, METHOD.POST)` is called like in line 220, solrResponse.getResults() is just initialized as an empty list. 
You can see in line 223  that the same thing happens and then the results get looped through.
